src/mds/Locker.cc

```diff
@@ -3633,10 +3633,21 @@ void Locker::_do_snap_update(CInode *in, snapid_t snap, int dirty, snapid_t foll
   // xattr
   if (xattrs) {
     dout(7) << " xattrs v" << i->xattr_version << " -> " << m->head.xattr_version
-	    << " len " << m->xattrbl.length() << dendl;
-    i->xattr_version = m->head.xattr_version;
-    auto p = m->xattrbl.cbegin();
-    decode(*px, p);
+            << " len " << m->xattrbl.length() << dendl;
+    __u32 total = 0;
+    for (const auto &p : *px)
+      total += (p.first.length() + p.second.length());
+    if ((total + m->xattrbl.length()) > mds->mdsmap->get_max_xattr_size()) {
```

Comment:
Shouldn't this be `total > mds->mdsmap->get_max_attr_size()`? IOW, why include the bufferlist length?

Reply:
Argh!  I'll need to go re-read this code again.  I **think** you may have a point, but probably what this should be is `m->xattrbl.length()) > mds->mdsmap->get_max_xattr_size()` because that's what the new length will be.  But I'll need to revisit this.  I'll try to have a look later today.

Reply:
I think this change is fine. `total` points to the total length of the xattrs that is already a part of the inode and `m->xattrbl.length()` is the length that the client is wanting to update. So, adding them up should not exceed the max size.

I probably confused you, however, the total calculation in the `for` loop is probably what confused me. So, that needs to wrapped inside `{}` :)

Reply:
In fact there's a real problem with this code, which I'm currently trying to fix:

`total` contains the total size of the xattrs pairs (names+values lengths), while `m->xattrbl.length()` has something completely different: it has to total size of the xattrs blob sent by the client, which includes **all** the xattrs (both the "old" ones and any "new" ones that the client may have buffered).

So, if we really want to know the total size of the new xattrs map, we need to decode `m->xattrbl` and compute the length of each key/value.

I plan to update this PR with this fixed soon, I'm currently working on it.

Reply:
@luis-henrix yeh - its not the first time that this has been the cause of confusion. Thx for clarifying it.